src/algebra/lie/cartan_subalgebra.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.lie.nilpotent
+
+/-!
+# Cartan subalgebras
+
+Cartan subalgebras are one of the most important concepts in Lie theory. We define them here.
+The standard example is the set of diagonal matrices in the Lie algebra of matrices.
+
+## Main definitions
+
+  * `lie_subalgebra.normalizer`
+  * `lie_subalgebra.le_normalizer_of_exists_ideal`
+  * `lie_subalgebra.is_cartan_subalgebra`
+
+## Tags
+
+lie subalgebra, normalizer, idealizer, cartan subalgebra
+-/
+
+universes u v w w₁ w₂
+
+variables {R : Type u} {L : Type v}
+variables [comm_ring R] [lie_ring L] [lie_algebra R L] (H : lie_subalgebra R L)
+
+namespace lie_subalgebra
+
+/-- The normalizer of a Lie subalgebra `H` is the set of elements of the Lie algebra whose product
+with any element of `H` lies in `H`. It is the Lie algebra equivalent of the group-theoretic
+normalizer (see `subgroup.normalizer`) and is an idealizer in the sense of abstract algebra. -/
+def normalizer : lie_subalgebra R L :=
+{ carrier   := { x : L | ∀ (y : L), (y ∈ H) → ⁅x, y⁆ ∈ H },
+  zero_mem' := λ y hy, by { rw zero_lie y, exact H.zero_mem, },
+  add_mem'  := λ z₁ z₂ h₁ h₂ y hy, by { rw add_lie, exact H.add_mem (h₁ y hy) (h₂ y hy), },
+  smul_mem' := λ t y hy z hz, by { rw smul_lie, exact H.smul_mem t (hy z hz), },
+  lie_mem'  := λ z₁ z₂ h₁ h₂ y hy, by
+    { rw lie_lie, exact H.sub_mem (h₁ _ (h₂ y hy)) (h₂ _ (h₁ y hy)), }, }
+
+lemma mem_normalizer_iff (x : L) : x ∈ H.normalizer ↔ ∀ (y : L), (y ∈ H) → ⁅x, y⁆ ∈ H := iff.rfl
+
+lemma le_normalizer : H ≤ H.normalizer :=
+begin
+  rw le_def, intros x hx,
+  simp only [submodule.mem_coe, mem_coe_submodule, coe_coe, mem_normalizer_iff] at ⊢ hx,
+  intros y, exact H.lie_mem hx,
+end
+
+/-- A Lie subalgebra is an ideal of its normalizer. -/
+lemma ideal_in_normalizer : ∃ (I : lie_ideal R H.normalizer), ↑I = of_le H.le_normalizer :=
+begin
+  simp only [exists_nested_lie_ideal_coe_eq_iff, mem_normalizer_iff],
+  intros x y h hy, exact h y hy,
+end
+
+/-- The normalizer of a Lie subalgebra `H` is the maximal Lie subalgebra in which `H` is a Lie
+ideal. -/
+lemma le_normalizer_of_exists_ideal {N : lie_subalgebra R L}
+  (h₁ : H ≤ N) (h₂ : ∃ (I : lie_ideal R N), ↑I = of_le h₁) : N ≤ H.normalizer :=
+begin
+  intros x hx,
+  rw mem_normalizer_iff,
+  intros y hy,
+  rw exists_nested_lie_ideal_coe_eq_iff at h₂,
```

Comment:
Does it make sense to just make `h2` equal to the result of this `rw` to begin with?

Reply:
Yeah I tend to agree. I experimented with various possible ways to express this (e.g., this [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/subobject.20predicates/near/227088411)) but what you suggest is probably best in the end since anyone using this lemma is bound to immediately invoke `exists_nested_lie_ideal_coe_eq_iff`!

Reply:
Now done.